enhancements/ingress/default-apps-domain.md

```diff
@@ -0,0 +1,118 @@
+approvers:
+  - "@knobunc"
+  - "@Miciah"
+creation-date: 2020-08-12
+last-updated: 2020-09-04
+status: implementable
+---
+
+# Default Apps Domain
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement allows for the configuration of a default apps domain in which
+can be used to set the default host and domain of the route, if not specified by
+the admin. Instead of using the default cluster domain, this new setting can be
+used to override it when set.
+
+## Motivation
+
+Many customers have requested "Custom Domains" in which they would like an already
+installed Openshift cluster to support additional domains for their apps by default.
+This domain is owned by a customer in which they would point a wildcard CNAME record
+to an existing well known CNAME record.
+
+### Goals
+
+1. Enable the cluster administor to specify the default apps domain seperate
+from the installed cluster domain.
+
+## Proposal
+
+The Ingress API is extended by adding an optional `AppsDomain` field with type
+`string`.
+
+```go
+type IngressSpec struct {
+    ...
+    // optional domain to be used instead of Domain installed cluster domain
+    // when Route Subdomain is unspecified.
+    // This assumes a new ingresscontroller has been setup with a wildcard
+    // certificate.
+    // +optional
+    AppsDomain string `json:"appsDomain,omitempty"`
+}
+```
+
+The following example configures apps.acme.io as the default apps domain.
+
+```yaml
+apiVersion: config.openshift.io/v1
+kind: Ingress
+metadata:
+  name: cluster
+spec:
+  domain: apps.drow-dev01.w9h5.s1.openshiftapps.com
+  appsDomain: apps.acme.io
+```
+
+### Implementation Details
+
+Implementation of this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/cluster-openshift-apiserver-operator
+
+In [config/v1/types_ingress.go](https://github.com/openshift/api/blob/master/config/v1/types_ingress.go) of [openshift/api](https://github.com/openshift/api), add new field: `appsDomain`.
+In [observe_ingresses.go](https://github.com/openshift/cluster-openshift-apiserver-operator/blob/master/pkg/operator/configobservation/ingresses/observe_ingresses.go) of [openshift/cluster-openshift-apiserver-operator](https://github.com/openshift/cluster-openshift-apiserver-operator), add logic to check if `appsDomain` is set. If that field is set use it insead of Spec.Domain for the value of `routingDomain`.
+
+### Risks and Mitigations
+
+Certain routes are created by various operators at install time. These routes
+could be potentially be deleted and re-added post-install time. Implementation
+must ensure these routes use the original domain when re-added.
```

Comment:
Does "Implementation" means whatever operator creates the route?  For example, cluster-foo-operator creates route "foo", and after this enhancement, cluster-foo-operator must ensure that it specifies a value for `spec.host` based on `domain` from the cluster ingress config when the operator creates route "foo".

Reply:
That's correct. The concern here is that if the route was created by an operator without a `spec.host` value, if the route is re-added it could take on the new `spec.appsDomain` value.

Reply:
@dustman9000 what do you propose the operators that utilize routes do to circumvent this? Console and auth come to mind...after this enhancement is implemented, should the console/auth operators be specifying `spec.host` on their routes?

Reply:
@sgreene570 - Correct, there is no need to move the console/oauth routes under the new domain. For now, since this enhancement is primarily for user-created routes, I would suggest we restrict this enhancement to user-created namespaces (ie. not in openshift-* and redhat-*).

Reply:
@sgreene570, note that setting `appsDomain` is day-2 (post-install) configuration.  This means that operators will have already created the routes, and the API will have set default values for `spec.host` based on the ingress domain if the operators did not specify `spec.host`.  The only concern is that the cluster administrator could modify `appsDomain` and something could subsequently delete the operator-created routes, and if the operators then recreated the routes without specifying an explicit `spec.host` value, they API would use the wrong domain (`appsDomain` instead of `domain`) for the operator-created routes.

Reply:
The authentication operator specifies `spec.host`:  https://github.com/openshift/cluster-authentication-operator/blob/b9d7a84f9db82d0e622b0d0b9c5d9162e810985f/pkg/controllers/metadata/metadata_controller.go#L245

The console operator does not:  https://github.com/openshift/console-operator/blob/b195cebf52ea7af30178f4704ac269f482d70208/pkg/console/subresource/route/route.go#L69-L85

@spadgett, to address this contingency, it might make sense for the console operator to do something similar to the authentication operator for the console's default route.